torch/distributed/fsdp/flat_param.py

```diff
@@ -466,65 +504,131 @@ def _init_flat_param(
+                    total_numel_without_padding += param.numel()
+        if len(params_to_flatten) == 0:
+            raise ValueError(
+                f"`params` were not found in `module`'s tree"
+                f"params: {params}\nmodule: {module}"
+            )
+        if (
+            self.rank == 0
+            and aligned_numel > 0
+            and total_numel != total_numel_without_padding
+        ):
+            log.info(
+                f"FSDP FlatParameter address alignment created "
+                f"{total_numel - total_numel_without_padding} "
+                f"numel of padding ({total_numel} vs. {total_numel_without_padding})"
+            )
+        # Pass `aligned_numel=0` since we already included padding tensors
+        self.flat_param: FlatParameter = self.flatten_tensors_into_flat_param(
+            params_to_flatten,
+            aligned_numel=0,
+            requires_grad=requires_grad,
         )
-        # For `use_orig_params=True`, ensure that the logical parameters are
-        # `nn.Parameter`s (and not plain `torch.Tensor`)
-
-        def convert_to_params(
-            tensors: List[Union[torch.Tensor, nn.Parameter]]
-        ) -> List[nn.Parameter]:
-            return [
-                t if isinstance(t, nn.Parameter) else nn.Parameter(t) for t in tensors
-            ]
-
         self.flat_param._init_metadata(
             param_infos,
             numels,
             shapes,
             fqns,
             shared_param_infos,
             param_extensions,
-            convert_to_params(params_to_flatten) if use_orig_params else None,
-            convert_to_params(shared_params) if use_orig_params else None,
+            _convert_to_params(params_to_flatten) if use_orig_params else None,
+            _convert_to_params(shared_params) if use_orig_params else None,
         )
 
-    @staticmethod
-    def flatten_params(
-        params: Sequence[torch.Tensor],
-        requires_grad: bool,
-    ) -> FlatParameter:
+    def _validate_tensors_to_flatten(
+        self, tensors: List[Union[Tensor, nn.Parameter]]
+    ) -> Tuple:
+        """
+        Validates the tensors to flatten and returns any necessary metadata.
+        """
+        dtype: Optional[torch.dtype] = None
+        requires_grad: Optional[bool] = None
+        device: Optional[torch.device] = None
+        for tensor in tensors:
+            if type(tensor) is FlatParameter:
+                raise ValueError("Cannot flatten a `FlatParameter`")
+            if dtype is None and not tensor.is_floating_point():
+                raise ValueError("Cannot flatten integer dtype tensors")
+            if dtype is not None and tensor.dtype != dtype:
+                raise ValueError(
+                    f"Must flatten tensors with uniform dtype but got {dtype} "
+                    f"and {tensor.dtype}"
+                )
+            # TODO: Relax the following for `use_orig_params=True`.
```

Comment:
does this mean use_orig_params=True does not technically support different requires_grad within a flat param currently? can we file an issue?

Reply:
https://github.com/pytorch/pytorch/issues/91167